ngmix/em.py

```diff
@@ -1,222 +1,1549 @@
-            pars = gm.get_full_pars()
-            self._gm = GMix(pars=pars)
+_em_conf_dtype = [
+    ('tol', 'f8'),
+    ('maxiter', 'i4'),
+    ('miniter', 'i4'),
+    ('sky', 'f8'),
+    ('vary_sky', 'bool'),
+    ('pixel_scale', 'f8'),
+]
+_em_conf_dtype = np.dtype(_em_conf_dtype, align=True)
 
-            if numiter >= maxiter:
-                flags = EM_MAXITER
+_sums_dtype = [
+    ('gi', 'f8'),
 
-            result = {
-                "flags": flags,
-                "numiter": numiter,
-                "fdiff": fdiff,
-                "message": "OK",
-            }
+    # used for convergence tests only
+    ('logtau', 'f8'),
+    ('logdet', 'f8'),
 
-        except (GMixRangeError, ZeroDivisionError) as err:
-            # most likely the algorithm reached an invalid gaussian
-            message = str(err)
-            print(message)
-            result = {
-                "flags": EM_RANGE_ERROR,
-                "message": message,
-            }
+    # scratch on a given pixel
+    ('tvsum', 'f8'),
+    ('tusum', 'f8'),
 
-        self._result = result
+    ('tu2sum', 'f8'),
+    ('tuvsum', 'f8'),
+    ('tv2sum', 'f8'),
 
-    run_em = go
+    # sums over all pixels
+    ('pnew', 'f8'),
+    ('vsum', 'f8'),
+    ('usum', 'f8'),
+    ('u2sum', 'f8'),
+    ('uvsum', 'f8'),
+    ('v2sum', 'f8'),
+]
+_sums_dtype = np.dtype(_sums_dtype, align=True)
 
-    def _make_sums(self, ngauss):
-        """
-        make the sum structure
-        """
-        return numpy.zeros(ngauss, dtype=_sums_dtype)
 
-    def _make_conf(self):
-        """
-        make the sum structure
-        """
-        conf_arr = numpy.zeros(1, dtype=_em_conf_dtype)
-        return conf_arr[0]
+_sums_dtype_fixcen = [
+    ('gi', 'f8'),
 
+    # used for convergence tests only
+    ('logtau', 'f8'),
+    ('logdet', 'f8'),
 
-_sums_dtype = [
-    ("gi", "f8"),
     # scratch on a given pixel
-    ("trowsum", "f8"),
-    ("tcolsum", "f8"),
-    ("tu2sum", "f8"),
-    ("tuvsum", "f8"),
-    ("tv2sum", "f8"),
+    ('tu2sum', 'f8'),
+    ('tuvsum', 'f8'),
+    ('tv2sum', 'f8'),
+
     # sums over all pixels
-    ("pnew", "f8"),
-    ("rowsum", "f8"),
-    ("colsum", "f8"),
-    ("u2sum", "f8"),
-    ("uvsum", "f8"),
-    ("v2sum", "f8"),
+    ('pnew', 'f8'),
+    ('u2sum', 'f8'),
+    ('uvsum', 'f8'),
+    ('v2sum', 'f8'),
 ]
-_sums_dtype = numpy.dtype(_sums_dtype, align=True)
+_sums_dtype_fixcen = np.dtype(_sums_dtype_fixcen, align=True)
 
-_em_conf_dtype = [
-    ("tol", "f8"),
-    ("maxiter", "i4"),
-    ("sky_guess", "f8"),
-    ("counts", "f8"),
-    ("pixel_scale", "f8"),
+_tau_dtype = [
+    ('logtau', 'f8'),
+    ('logdet', 'f8'),
+]
+_tau_dtype = np.dtype(_tau_dtype)
+
+_sums_dtype_fluxonly = [
+    ('gi', 'f8'),
+
+    # sums over all pixels
+    ('pnew', 'f8'),
 ]
-_em_conf_dtype = numpy.dtype(_em_conf_dtype, align=True)
+_sums_dtype_fluxonly = np.dtype(_sums_dtype_fluxonly, align=True)
 
 
 def test_1gauss(
```

Comment:
I had kept them because the provide visualization of the results, which the unit tests do not.

But maybe we an move these under and `examples/` directory along with some other examples